src/expand.rs

```diff
@@ -0,0 +1,140 @@
+use crate::parse::{parse_decimal, parse_id};
+use crate::{Captures, Error, Result};
+
+/// A set of options for expanding a template string using the contents
+/// of capture groups.  Create using the `builder` method.
+#[derive(Debug)]
+pub struct Expander<'a> {
```

Comment:
I'm not sure if this should be public API or not. Do you have a use case where you need other delimiters than the ones in `expand` and `expand_backslash`? If not, I'd probably keep this private for now.

Reply:
I'm not sure either, TBH. I _sort of_ have a use case. Basically I've noticed that while developers mostly treat regex syntax and the associated replacement/expansion syntax as implementation details, they end up being exposed to end users pretty often in things like text editors and command-line tools. My `expand_backslash` method is there because I have a Python program that exposes the syntax as part of its UI, and I wanted to port it to Rust without changing the UI.

I think the basic `expand` method is important to make `fancy_regex` a more complete drop-in replacement for `regex`, and I think either `replace_backslash` should exist, or `Expander` should be public, but probably not both. I like the idea of having an `Expander` type to accommodate multiple syntaxes, but the builder interface is probably overkill, so perhaps it should just have factory methods like `Expander::default()`, `Expander::python()`, etc.

Another nice thing about having an `Expander` type is that it's a place where you can put more expansion methods. The fact that `Captures::expand` has to append to an existing `String` makes it clunky for common use cases. That signature needs to be supported for compatibility with the `regex` create, and I can see why it would sometimes be preferred for performance, but I think most users would prefer a method that returns a new `String`.

I uploaded a new commit that adds the factory methods and changes the expansion methods a bit. If you like that design, I think I'll remove `expand_blackslash` and `ExpanderBuilder`, and update the tests accordingly.

Reply:
Thanks for that, that makes sense to me :)!

My initial reaction was that if we have `expand_backslash` we don't need `Expander::python`. But we do, because a user might want to use it with `io::Write`, not `&mut String`. If we get rid of `expand_backslash`, the user can't do `&mut String` without the complicated `Cursor` conversion. But having both still doesn't sit right with me (and also that their naming is not consistent).

How about these (small) changes:

1. On `Captures`, keep only `expand`, for regex compatibility. Add a note to the docs there to see `Expander` for others usages.
2. On `Expander`, add an additional method that can append to a `&mut String`. That then covers all possible append use cases in the "advanced" API.
3. I'm not sure about the naming of the 3 expand methods on `Expander` yet. Some ideas:
   1. `expand -> String`, `expand_to(&mut String)` (or `expand_append(&mut String)`) and `expand_write(impl io::Write)`
   2. `expanded -> String`, `expand(&mut String)` and `expand_to(impl io::Write)`

I think I like the first one. What do you think?